src/subdivide_multi.h

```diff
@@ -0,0 +1,302 @@
+/// @param[in] threshold            The user-input epsilon value for the distance check.
+/// @param[out] active          A `bool` represents whether the tet is containing part of the geometry, i.e., this tet passes the zero-crossing test.
+///
+/// @return         A `bool` represents whether the tet is "refinable", i.e., passing the zero-crossing test and contains error greater than `threshold`.
+///
+///
+bool subMI(const std::array<std::array<double, 3>,4> &pts,
+           const llvm_vecsmall::SmallVector<std::array<double,4>, 20> &vals,
+           const llvm_vecsmall::SmallVector<std::array<std::array<double, 3>,4>, 20> &grads,
+           const double threshold,
+           bool& active);
+
+/// Enums for the current settings of implicit complexes
+enum geo_obj {
+    IA,
+    CSG,
+    MI
+};
+
+/// Enums for CSG boolean operations
+enum csg_operations{
+    Intersection,
+    Union,
+    Negation
+};
+
+
+///Defines the enumeration of CSG operations. In the data structure, Intersection is 0, Union is 1, and Negation is 2.
+struct csg_unit{
+    int operation;
+    std::array<int, 2> elements;
+};
+
+///load the csg file
+///
+///@param[in] filename          The name of the input CSG tree
+///@param[out] tree         The loaded tree structure
+///
+bool load_csgTree(const std::string filename, llvm_vecsmall::SmallVector<csg_unit, 20>& tree);
+
+/// an iterative algortihm that traverses through the csg tree
+///
+///@param[in] csgTree           The CSG structure: a list of csg units
+///@param[in] curNode           The current index in the csg structure
+///@param[in] funcInt           The intervals of all the functions
+///
+///@param[out] std::pair
+std::pair<std::array<double, 2>, llvm_vecsmall::SmallVector<int, 20>> iterTree(const llvm_vecsmall::SmallVector<csg_unit, 20>csgTree,const int curNode,const llvm_vecsmall::SmallVector<std::array<double , 2>, 20> funcInt);
+
+/// Below are the variable/constant/function definitions that will only be used in `subdivide_multi.cpp`
+///
+///
+/// Stores the index of permutations of n less than `funcNum` for the use in `subTet` and `subMI` functions.
+///
+extern llvm_vecsmall::SmallVector<llvm_vecsmall::SmallVector<llvm_vecsmall::SmallVector<std::array<int, 4>, 100>, 3>, 20> multiple_indices;
+
+/// Definations of varialbes before the computations to avoid repetitive defnitions.
+extern int sub_call_two;
+extern int sub_call_three;
+extern int GLOBAL_METHOD;
+extern bool curve_network;
+extern llvm_vecsmall::SmallVector<csg_unit, 20> GLOBAL_CSGTREE;
+extern llvm_vecsmall::SmallVector<std::array<double, 20>, 20> valList;
+extern llvm_vecsmall::SmallVector<std::array<double, 16>, 20> diffList;
+extern llvm_vecsmall::SmallVector<bool, 20> activeTF;
+extern llvm_vecsmall::SmallVector<std::array<double , 2>, 20> funcInt;
+extern llvm_vecsmall::SmallVector<std::array<double, 3>, 20> gradList;
+extern size_t funcNum;
```

Comment:
Why are these global variables necessary? The subdivision checks should be embarrassingly parallel (ideally run in parallel), there should be no shared data/state between different checks.

Reply:
If the only purpose for these global variables is to avoid repeated memory allocation in sequential computation, they should not be in the global scope. Using local variable with some book-keeping should suffice.  In fact, there should be no global variables in our code (see [here for reasons](https://luzkan.github.io/smells/global-data)).